### PR TITLE
Copter: add disarm aux switch RCn_OPTION=81

### PR DESCRIPTION
**Pull request by tatsuy**
This is similar to Estop and Arm/Disarm, but there is a slight difference.

When you set the Disarm switch to High, the copter immediately disarms.
Arming will fail if the Disarm switch is enabled.

Emergency Stop is not immediately disarmed (the timing is depends on DISARM_DELAY).
Without noticing that it has not been disarmed, it is possible to return the switch to Low and unintentionally start the motor running again.

With the Arm/Disarm option, there is a risk that the switch will be inadvertently set to High and start to rotate the motors.

### CODE REVIEW

**Comment by muramura** [comment]
I don't think this feature is necessary.

I've PR'd in the past about the delay time for the emergency stop feature.
I enabled the delay time 0.
However, we have received comments that this emergency stop function will stop the motor in the primary to recover the sensor abnormal condition in flight.
We received a comment that it is better to use the ARM/DISARM function when you want to stop the motor immediately.
When using the ARM/DISARM switch, the ARM mode is set to HIGH.
If you set it to LOW by mistake, it would be better to set it to HIGH immediately and to ARM.

Translated with www.DeepL.com/Translator (free version)

**Comment by tatsuy** [comment]
@muramura 
Thank you for your comment.
It's important that you wanted the same functionality in the past.

Modifying existing functionality is not a good idea.
This change does not have a risk as it adds a new option.

File: ArduCopter/Copter.h [hunk 1]
```diff
@@ -368,6 +368,7 @@ class Copter : public AP_Vehicle {
             uint8_t unused3                 : 1; // 25      // was compass_init_location; true when the compass's initial location has been set
             uint8_t unused2                 : 1; // 26      // aux switch rc_override is allowed
             uint8_t armed_with_switch       : 1; // 27      // we armed using a arming switch
+            uint8_t disarm_switch           : 1; // 28      // true if disarm aux switch is enable
```
Comment (peterbarker) [inline ArduCopter/Copter.h:371]: Do we need this state?

Can arming just check the channel `aux_switch_pos` instead?

File: ArduCopter/RC_Channel.cpp [hunk 2]
```diff
@@ -111,6 +111,7 @@ void RC_Channel_Copter::init_aux_function(const aux_func_t ch_option, const aux_
     case AUX_FUNC::SUPERSIMPLE_MODE:
     case AUX_FUNC::SURFACE_TRACKING:
     case AUX_FUNC::WINCH_ENABLE:
+    case AUX_FUNC::DISARM:
```
Comment (peterbarker) [inline ArduCopter/RC_Channel.cpp:114]: Everything here is in alphabetical order.  We ought to not mess that up.

File: ArduCopter/AP_Arming.cpp [hunk 3]
```diff
@@ -633,7 +633,13 @@ bool AP_Arming_Copter::arm_checks(AP_Arming::Method method)
         SRV_Channels::set_emergency_stop(false);
         // if we are using motor Estop switch, it must not be in Estop position
     } else if (rc().find_channel_for_option(RC_Channel::AUX_FUNC::MOTOR_ESTOP) && SRV_Channels::get_emergency_stop()){
-        gcs().send_text(MAV_SEVERITY_CRITICAL,"Arm: Motor Emergency Stopped");
```
Comment (peterbarker) [inline ArduCopter/AP_Arming.cpp:636]: Good, unrelated, change.  Should be in a separate PR to simplify this one.

File: ArduCopter/AP_Arming.cpp [hunk 4]
```diff
@@ -633,7 +633,13 @@ bool AP_Arming_Copter::arm_checks(AP_Arming::Method method)
...
+        check_failed(true, "Motor Emergency Stopped");
+        return false;
+    }
+
+    // if we are not using Disarm switch and it's enabled, fail to arm
```
Comment (peterbarker) [inline ArduCopter/AP_Arming.cpp:640]:
```suggestion
    // if we are using Disarm switch and it's asserted, fail to arm
```

File: ArduCopter/RC_Channel.cpp [hunk 5]
```diff
@@ -561,6 +562,21 @@ void RC_Channel_Copter::do_aux_function(const aux_func_t ch_option, const aux_sw
             }
             break;
             
+        case AUX_FUNC::DISARM:
+            switch (ch_flag) {
+            case HIGH:
+                AP::arming().disarm(AP_Arming::Method::AUXSWITCH);
```
Comment (peterbarker) [inline ArduCopter/RC_Channel.cpp:568]: We could consider adding another `Method` to disambiguate, but I don't think it's required.

File: ArduCopter/RC_Channel.cpp (hunk 2 above)
Comment (tatsuy) [reply ArduCopter/RC_Channel.cpp]: fixed it!

File: ArduCopter/AP_Arming.cpp (hunk 3 above)
Comment (tatsuy) [reply ArduCopter/AP_Arming.cpp:636]: fixed it!

File: ArduCopter/AP_Arming.cpp (hunk 4 above)
Comment (tatsuy) [reply ArduCopter/AP_Arming.cpp]: fixed it!

File: ArduCopter/Copter.h (hunk 1 above)
Comment (tatsuy) [reply ArduCopter/Copter.h]: I've fixed to use aux_switch_pos_t.

File: ArduCopter/AP_Arming.cpp (hunk 3 above)
Comment (peterbarker) [reply ArduCopter/AP_Arming.cpp:636]: Not seeing that PR.

File: ArduCopter/AP_Arming.cpp [hunk 6]
```diff
@@ -633,7 +633,14 @@ bool AP_Arming_Copter::arm_checks(AP_Arming::Method method)
...
         // if we are using motor Estop switch, it must not be in Estop position
     } else if (rc().find_channel_for_option(RC_Channel::AUX_FUNC::MOTOR_ESTOP) && SRV_Channels::get_emergency_stop()){
-        gcs().send_text(MAV_SEVERITY_CRITICAL,"Arm: Motor Emergency Stopped");
+        check_failed(true, "Motor Emergency Stopped");
+        return false;
+    }
+
+    // if we are using Disarm switch and it's asserted, fail to arm
+    RC_Channel *chan = rc().find_channel_for_option(RC_Channel::AUX_FUNC::DISARM);
+    if (chan != nullptr && chan->get_aux_switch_pos() == RC_Channel::aux_switch_pos_t::HIGH){
```
Comment (peterbarker) [inline ArduCopter/AP_Arming.cpp:642]: This code now appears to be duplicated in AP_Arming... should be removed here?

File: ArduPlane/AP_Arming.cpp [hunk 7]
```diff
@@ -104,6 +104,12 @@ bool AP_Arming_Plane::pre_arm_checks(bool display_failure)
         ret = false;
     }
 
+    RC_Channel *chan = rc().find_channel_for_option(RC_Channel::AUX_FUNC::DISARM);
+    if (chan != nullptr && chan->get_aux_switch_pos() == RC_Channel::aux_switch_pos_t::HIGH){
```
Comment (peterbarker) [inline ArduPlane/AP_Arming.cpp:108]: Should this be != LOW instead?

File: ArduPlane/AP_Arming.cpp [hunk 8]
```diff
@@ -104,6 +104,12 @@ bool AP_Arming_Plane::pre_arm_checks(bool display_failure)
         ret = false;
     }
 
+    RC_Channel *chan = rc().find_channel_for_option(RC_Channel::AUX_FUNC::DISARM);
```
Comment (peterbarker) [inline ArduPlane/AP_Arming.cpp:107]: `chan` can be const here.

File: ArduPlane/AP_Arming.cpp (hunk 7 above)
Comment (tatsuy) [reply ArduPlane/AP_Arming.cpp]: Disarm is triggered when HIGH.
!= LOW means arming fails by MIDDLE position.
I think it's better to unify with HIGH.

File: ArduPlane/AP_Arming.cpp (hunk 8 above)
Comment (tatsuy) [reply ArduPlane/AP_Arming.cpp]: fixed it

File: ArduCopter/AP_Arming.cpp (hunk 6 above)
Comment (tatsuy) [reply ArduCopter/AP_Arming.cpp]: Thank you!
I learned a lot.

File: ArduCopter/AP_Arming.cpp (hunk 3 above)
Comment (tatsuy) [reply ArduCopter/AP_Arming.cpp]: I misread.
Done!

**Comment by tatsuy** [comment]
Tested on SITL with copter, rover and plane respectively.

File: libraries/AP_Arming/AP_Arming.h [hunk 9]
```diff
@@ -36,6 +36,7 @@ class AP_Arming {
         ARMING_CHECK_CAMERA      = (1U << 16),
         ARMING_CHECK_AUX_AUTH    = (1U << 17),
         ARMING_CHECK_VISION      = (1U << 18),
+        ARMING_CHECK_DISARM      = (1U << 19),
```
Comment (peterbarker) [inline libraries/AP_Arming/AP_Arming.h:39]: We probably don't want to set aside an arming bit for this.  Is there an existing bit which looks appropriate?

We could also consider the implications of just putting it behind arming-check-all.

File: libraries/RC_Channel/RC_Channel.cpp [hunk 10]
```diff
@@ -980,6 +987,28 @@ bool RC_Channel::read_3pos_switch(RC_Channel::aux_switch_pos_t &ret) const
     return true;
 }
 
+// return switch position value as LOW, MIDDLE, HIGH
```
Comment (peterbarker) [inline libraries/RC_Channel/RC_Channel.cpp:990]: This comment is rather incomplete considering the extra logic involved.  You need to explain that if reading the switch fails then it returns low.

File: libraries/RC_Channel/RC_Channel.cpp [hunk 11]
```diff
@@ -980,6 +987,28 @@ bool RC_Channel::read_3pos_switch(RC_Channel::aux_switch_pos_t &ret) const
...
+RC_Channel::aux_switch_pos_t RC_Channel::get_aux_switch_pos() const
+{
+    RC_Channel::aux_switch_pos_t position = RC_Channel::LOW;
+    if (!read_3pos_switch(position)) {
```
Comment (peterbarker) [inline libraries/RC_Channel/RC_Channel.cpp:994]: Ignore the result using `UNUSED_RESULT` and don't bother with the `if`.  It's obvious enough what's happening.

File: libraries/RC_Channel/RC_Channel.cpp [hunk 12]
```diff
@@ -980,6 +987,28 @@ bool RC_Channel::read_3pos_switch(RC_Channel::aux_switch_pos_t &ret) const
...
+        return RC_Channel::LOW;
+    }
+
+    return position;
+}
+
+// return switch position value as LOW, MIDDLE, HIGH
```
Comment (peterbarker) [inline libraries/RC_Channel/RC_Channel.cpp:1001]: Again, special logic requires more comments - here and in the header.

File: libraries/AP_Arming/AP_Arming.cpp [hunk 13]
```diff
@@ -1216,6 +1217,18 @@ bool AP_Arming::visodom_checks(bool display_failure) const
     return true;
 }
 
+// check disarm switch is asserted
+bool AP_Arming::disarm_switch_checks(bool display_failure) const
+{
+    const RC_Channel *chan = rc().find_channel_for_option(RC_Channel::AUX_FUNC::DISARM);
+    if (chan != nullptr && chan->get_aux_switch_pos() == RC_Channel::aux_switch_pos_t::HIGH){
```
Comment (peterbarker) [inline libraries/AP_Arming/AP_Arming.cpp:1224]: Should this be checking `!= RC_Channel::aux_switch_pos_t::LOW`?

Comment (tatsuy) [reply libraries/AP_Arming/AP_Arming.cpp]: I think it's OK to be able to arm in the MIDDLE position.
https://github.com/ArduPilot/ardupilot/pull/14244#discussion_r418537445

File: libraries/AP_Arming/AP_Arming.h (hunk 9 above)
Comment (tatsuy) [reply libraries/AP_Arming/AP_Arming.h]: I couldn't find a suitable one.
But, I fixed about ARMING_CHECK_ALL.

File: libraries/RC_Channel/RC_Channel.cpp (hunk 10 above)
Comment (tatsuy) [reply libraries/RC_Channel/RC_Channel.cpp:990]: thank you, I fixed it.

File: libraries/RC_Channel/RC_Channel.cpp (hunk 11 above)
Comment (tatsuy) [reply libraries/RC_Channel/RC_Channel.cpp]: fixed it.

**Comment by tatsuy** [comment]
@peterbarker 
Thank you for the review!
I referenced this code for getting aux switch positions.
https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_Camera/AP_RunCam.cpp#L269

**Comment by andyp1per** [comment]
> @peterbarker
> Thank you for the review!
> I referenced this code for getting aux switch positions.
> https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_Camera/AP_RunCam.cpp#L269

I think this is generically useful enough that it should live in RCChannels

**Comment by peterbarker** [comment]
@andyp1per So this PR doesn't *move*, it copies - should it be moved?

File: libraries/AP_Arming/AP_Arming.cpp (hunk 13 above)
Comment (peterbarker) [reply libraries/AP_Arming/AP_Arming.cpp]: I think I'm going to disagree on this one, but that's not a blocker.  Other devs may want to chime in.

Leaving the middle switch position "reserved for future use" won't cost us anything at this point, and it makes for a cleaner story to restrict this to just the two positions (`HIGH`/`LOW`).

File: libraries/AP_Arming/AP_Arming.cpp [hunk 14]
```diff
@@ -1216,6 +1217,20 @@ bool AP_Arming::visodom_checks(bool display_failure) const
     return true;
 }
 
+// check disarm switch is asserted
+bool AP_Arming::disarm_switch_checks(bool display_failure) const
+{
+    const RC_Channel *chan = rc().find_channel_for_option(RC_Channel::AUX_FUNC::DISARM);
+    if (chan != nullptr &&
+            chan->get_aux_switch_pos() == RC_Channel::aux_switch_pos_t::HIGH &&
```
Comment (peterbarker) [inline libraries/AP_Arming/AP_Arming.cpp:1225]: Whitepsace issues; indented incorrectly.

File: libraries/RC_Channel/RC_Channel.cpp [hunk 15]
```diff
@@ -980,6 +987,16 @@ bool RC_Channel::read_3pos_switch(RC_Channel::aux_switch_pos_t &ret) const
     return true;
 }
 
+// return switch position value as LOW, MIDDLE, HIGH
+// if reading the switch fails then it returns LOW
+RC_Channel::aux_switch_pos_t RC_Channel::get_aux_switch_pos() const
+{
+    RC_Channel::aux_switch_pos_t position = RC_Channel::LOW;
```
Comment (peterbarker) [inline libraries/RC_Channel/RC_Channel.cpp:994]:
```suggestion
    RC_Channel::aux_switch_pos_t position = aux_switch_pos_t::LOW;
```

File: libraries/AP_Arming/AP_Arming.cpp (hunk 14 above)
Comment (tatsuy) [reply libraries/AP_Arming/AP_Arming.cpp]: fixed it

File: libraries/RC_Channel/RC_Channel.cpp (hunk 15 above)
Comment (tatsuy) [reply libraries/RC_Channel/RC_Channel.cpp]: fixed it

File: libraries/RC_Channel/RC_Channel.cpp [hunk 16]
```diff
@@ -980,6 +987,16 @@ bool RC_Channel::read_3pos_switch(RC_Channel::aux_switch_pos_t &ret) const
     return true;
 }
 
+// return switch position value as LOW, MIDDLE, HIGH
+// if reading the switch fails then it returns LOW
+RC_Channel::aux_switch_pos_t RC_Channel::get_aux_switch_pos() const
+{
+    RC_Channel::aux_switch_pos_t position = aux_switch_pos_t::LOW;
```
Comment (peterbarker) [inline libraries/RC_Channel/RC_Channel.cpp:994]:
```suggestion
    aux_switch_pos_t position = aux_switch_pos_t::LOW;
```

This is a niggle and doesn't really need to be fixed.

Comment (tatsuy) [reply libraries/RC_Channel/RC_Channel.cpp]: Thank you! 
It should be fixed.

**Comment by andyp1per** [comment]
> 
> 
> @andyp1per So this PR doesn't _move_, it copies - should it be moved?

I'm using it in my OSD PR - happy to move it there or here

**Comment by tatsuy** [comment]
@andyp1per 
I added your function to RC_Channels.
I confirmed that the function works.
Please double check.

@peterbarker 
Let me know if I need to separate PR.

File: libraries/RC_Channel/RC_Channel.cpp [hunk 17]
```diff
@@ -980,6 +987,24 @@ bool RC_Channel::read_3pos_switch(RC_Channel::aux_switch_pos_t &ret) const
...
+    UNUSED_RESULT(read_3pos_switch(position));
+
+    return position;
+}
+
+// return switch position value as LOW, MIDDLE, HIGH
+// if reading the switch fails then it returns LOW
+RC_Channel::aux_switch_pos_t RC_Channels::get_channel_pos(const uint8_t rcmapchan) const
+{
+    const RC_Channel* chan = rc().channel(rcmapchan-1);
```
Comment (andyp1per) [inline libraries/RC_Channel/RC_Channel.cpp:1004]: You need to check for nullptr here

Comment (tatsuy) [reply libraries/RC_Channel/RC_Channel.cpp:1004]: thank you! fixed it.

File: libraries/RC_Channel/RC_Channel.cpp [hunk 18]
```diff
@@ -980,6 +987,24 @@ bool RC_Channel::read_3pos_switch(RC_Channel::aux_switch_pos_t &ret) const
...
+RC_Channel::aux_switch_pos_t RC_Channel::get_aux_switch_pos() const
+{
+    aux_switch_pos_t position = aux_switch_pos_t::LOW;
+    UNUSED_RESULT(read_3pos_switch(position));
+
+    return position;
+}
+
+// return switch position value as LOW, MIDDLE, HIGH
+// if reading the switch fails then it returns LOW
```
Comment (peterbarker) [inline libraries/RC_Channel/RC_Channel.cpp:1001]: This comment should also be in the header

**Comment by rmackay9** [comment]
Merged, thanks!

**Comment by tatsuy** [comment]
@rmackay9 
Thank you for the merging!

@peterbarker 
Thank you so much for your review.
It helped me to improve my knowledge.
